### ISSUE TITLE
Filtered getContext? or other solution?

### ISSUE THREAD

**Issue by johngid** [IBM-i2/Analyze#45]
I haven't seen the option of a "Filtered getContext" like when expanding from the AR. I am in a situation where the second level of information (a getContext) would bring too much information about the entity. It would take around 15 - 40 minutes depending on how many entities are related to my selected entity.

I am trying to workaround this, by doing several DAODs which would mimic this behaviour. However in ANBP, when you chart an item sent from the Intelligence Portal, the Database Identifier only shows me the DAOD the item is comming from. When I search each DAOD for the same entity, and manually merge them in ANBP, I have the result I want, and I can select from which DAOD I want to expand.

I am looking for a way to have this merger automatically, or in other words, to have a single search, and expand to any of my DAODs.

How can I achieve this?

**Comment by TonyJon** [comment]
Hi John

What you want to do can be done as long as you do not have too many items to retrieve  but it is quite involved and may not be viable if you actually have a lot of items to retrieve even when filtered.

We do not have a filtered  "Filtered getContext" for DAOD as you have discopvered.

You say  that a your second getContext would bring back too much infomation about the entity. In actual fact we have a limit on the number of items that can be returned  when using getContext with a DAOD connector. This limit is configurable but if you increase it you should test that you can get these extra results back without causing an issue to your system. If you have not already seen it, there is a tech note about this here : http://www-01.ibm.com/support/docview.wss?uid=swg21979467 

Have you hit this pre configured limit ?

If you are on 4.1.3 of i2 Analyze or above we altered our naming to use Opal and Onyx in our folder structures so you will find  ApolloClientSettings.xml here:
C:\IBM\i2Analyze\toolkit\configuration\fragments\onyx-services-ar

or here:

C:\IBM\i2EIA\toolkit\configuration\fragments\onyx-services-ar
if you are deloying EIA.


The number of items that you are allowed to return from a getContext is passed to the DAOD connector as you can see from the JavaDoc for the IExternalDataItemRetriever here: 

https://ibm-i2.github.io/Analyze/docs/com/i2group/apollo/externaldata/edrs/IExternalDataItemRetriever.html

It is the responsibility of the DAOD connector to respect this maximum.

Your workaround to not having a "Filtered getContext" is an interesting one, I assume that each DAOD connector is written to only return dumbbells that link to a specific entity type  from your getContext  seed item.


When you configured  your DAOD connectors to enable showContext, one of the things you did was to set this property: 

ExternalDataCreateStandardEDRSRetrievalBlock=true

This is a good thing to do in the simple case where you have one connector for your item as it allows our core DAOD code to automatically generate a RetrievalBlock  based on the information you supplied for the item in the XML Source when you originally sent it to us.

The retrieval block tells us what DAOD connector you retrieved the item from and allows ANB to wire things up so that you can see the connector on the right click menu and run Show Context against it.

The problem in your case then is that when you chart the item from each of your filtering connectors it only has the one automatic retrieval block that references back to the item via the DAOD connector that it was retrieved through.

As these are all different you do not get an automatic merge in ANB.

What you need to do is make each of your DAOD connectors add the extra retrieval blocks for all of the  other DAOD connectors including your original one that you searched on.

At this point you would probably want to turn off our automatic generation of the single retrieval block for your DAOD connectors  you add in from, as you can then use the same retrieval block generation code in all of them.

If you manage to do this successfully then when you add your initial search results to ANB  from the i2 Analyze Portal, you will get a right click context menu entry on your items  for all of your filter DAOD  connectors and your original one.

Because each connector has the retrieval blocks for all other connectors,  the ends will merge on the ANB chart when you return results from any so you should not see duplicates in ANB.

In order for you to be able to get this to work  you will need to create Provenance information including the required retrieval blocks  in the XML source  you send back to us that is an exact replica of what we would have created  automatically. If you do not get this exactly right then we may not link back to your connectors correctly or may add extra  retrieval blocks of our own when you  run show context which will confuse things.

As you can see, this is not a small undertaking.

The CardProvenance and RetrievalBlock definition for our XML Source can be found in schema1.xsd in the schema mapping jar , one of the  XSD's that  is created  when you generate the mapping jar for our schema.

The easiest way to see what data values to put in to this is to  use Fiddler to  record what we are currently sending back  to the Client when you get results back from your connectors. The XML structure we send to ANBP or the i2A Portal is more complex than the one you have to define for the XSD but will show you the DATA values to use for each connector. (Currently  these will be including an auto generated retrieval block for each connector  when  you  search on it).

Here is an example of that output when using Developer essentials examples to create the ability to access both the DAOD  file based example and the Document example from a single  item that is searched for in the  file based example: 

![processedxmlforclient](https://user-images.githubusercontent.com/17141806/31501003-fcc0ed7e-af60-11e7-90e9-550de1cc3c25.jpg)



What I did here was to take  XSLT that we use to map to our DAOD XML format for the file based example connector and I added an extra transform  to put the required retrieval block information for the Documents example. You can try this using Dev essentials yourself .

(This was just the quickest way for me to inject the new retrieval block in to the XML)

![xsltforretrievalblocks](https://user-images.githubusercontent.com/17141806/31501117-41879ab6-af61-11e7-8c37-fd093fc476f9.jpg)


In this particular case I still allowed the core DAOD code to create the original retrieval  blocks as I always used the same connector to chart from so they would not appear in the XML that the XSLT creates and never used that one for a show context, but in your case you might want to add those others in to.

- The important things are:
- That you have an EDRS retrieval block for each connector
- That you have an ID in the Retrieval block with a a type  that is what we expect for DAOD connectors
- That within the RetrievalData you have two RetrievalDatum elements  with keys for the Data source  (DAOD Connector) and the item itself.

![overview](https://user-images.githubusercontent.com/17141806/31502156-28107988-af64-11e7-9ae2-fb956582b3e5.jpg)

- That the key for the item has the same format and content  irrespective of which data source it comes from  it has to be the same for an automatic merge in ANB to happen.

I appreciate that there is a lot to think about here so I will leave this reply for now and  ask that you post back when you have had a chance to read/try it out and have more questions.

Cheers


Tony

**Comment by johngid** [comment]
What a great response. I appreciate it. I will try at first with the example files, and then with the actual implementation.

I really appreciate your help. I was amazed with the detail of this reponse. 

Thank you, and I'll keep you posted.

**Comment by TonyJon** [comment]
Hi John

There are just a couple of extra points that I wanted to highlight to save you having to figure out why things may not be working as you expect when you try this with the Developer Essentials examples.

The Developer Essentials examples  have XML validation switched on by default.
If you leave that switched on then you will need to populate more of the XML content than I have shown you to get this to pass validation.

You can tell if this  causing you an issue as you will get a  validation error like:

com.i2group.apollo.common.exception.ValidationRuntimeException: 
'cvc-complex-type.2.4.a: Invalid content was found starting with element 'RetrievalBlock'. One of '{"http://www.i2group.com/Schemas/2013-10-03/ModelData/Provenance":RetrievalBlock}' is expected.' at line -1, column -1.

You can switch this validation off  by altering the value for the  ValidateXml property in ApolloServerSettingsDaodMandatory.properties for each of the example connectors  to false and re-deploying.

Once you have done this and have replaced the GUID in the XSLT sample for the actual GUID of the document example in your environment, you can then test that if you send the full set of results to ANBP from the DAOD filesystem example you will get menu options for the document example as well .

You can also test that sending entities from either will not produce duplicates and will merge when appropriate.

You can, if you want,  augment these basic examples to support  show context but from the point of view of the exercise  that is not really necessary and once you have gotten to grips with the technique you may as well move on and apply it to your own connectors. 

Cheers

**Comment by johngid** [comment]
Hi TonyJon,

We were able to succesfully add the <Retrieval Block> of each DAOD (the filesystem-example, and the documents-example), and now the Entity automatically merges in ANBP.

We are on the right track now. We are now building a custom DAOD to test this, the information you provided was of a lot of help, and we got it working.

I will write our progress as we move forward.

I appreciate  your help.
